### PR TITLE
implement Filecoin.ChainSetHead endpoint

### LINKED ISSUE #2706: Implement Filecoin.ChainSetHead endpoint
**Issue summary**

Sometimes it's very useful to be able to reset the HEAD manually. Lotus supports it.

One use case would be for exporting a snapshot with Forest across upgrades.

E.g.

```
forest-cli --chain calibnet chain set-head 322351
forest-cli --chain calibnet snapshot export -t 322054 -r 2000
```

**Task summary**

- [ ] Add support for RPC-JSON `Filecoin.ChainSetHead` endpoint
- [ ] Add cli support in chain subcommand

**Acceptance Criteria**

- [ ] Setting HEAD to a previous epoch works and Forest can revalidate the chain section.

**Other information and links**

https://github.com/filecoin-project/lotus/blob/093ff9558000054e272fcb6457c54c9f8e0b8da1/api/api_full.go#L141

### PR DESCRIPTION
**Pull request by aatifsyed**
## Summary of changes

Changes introduced in this pull request:

- Add `ChainSetHead` endpoint
- Add `chain set-head` subcommand

### Live testing done:
- Run the local daemon, with an empty keystore passphrase. Let it sync.
```console
$ cargo run --bin forest -- --chain=calibnet --save-token=token
...
2023-05-04T01:26:01.742836Z  INFO forest_chain_sync::tipset_syncer: Successfully synced tipset range: [527905, 527906]
```
- Get a tipset from an earlier epoch using a block explorer. E.g https://calibration.filscan.io/tipset/chain?height=527666 gives us a tipset of `bafy2bzacecaqwjlwim5ylzcpkkktjvz6qz47mrrw5uaohu2nj7c4rnppohpds bafy2bzaced74lwq6emvroxrlcmblf64bo47p35da3immhvfhtfmkv6hd2jrpu`
- Run the new command
```console
$ cargo run --bin forest-cli -- --token=$(cat token) chain set-head bafy2bzacecaqwjlwim5ylzcpkkktjvz6qz47mrrw5uaohu2nj7c4rnppohpds bafy2bzaced74lwq6emvroxrlcmblf64bo47p35da3immhvfhtfmkv6hd2jrpu
```
- Observe that the local head is now behind, and that the daemon validates to catch up. (It may freeze//need restarting after running the command).
```console
$ cargo run --bin forest -- --chain=calibnet --save-token=token
...
2023-05-04T01:31:38.256067Z  INFO forest_chain_sync::chain_muxer: Local node is behind the network, starting BOOTSTRAP from LOCAL_HEAD = 527666 -> NETWORK_HEAD = 527917
...
2023-05-04T01:31:39.530603Z  INFO forest_chain_sync::tipset_syncer: Validating tipset: EPOCH = 527666, N blocks = 2
...
2023-05-04T01:32:22.487042Z  INFO forest_chain_sync::tipset_syncer: Validating tipset: EPOCH = 527667, N blocks = 4
```

## Reference issue to close (if applicable)

Closes #2706 

## Other information and links
I've got a low confidence in this change - I observed forest freezing up for O(minute) after running the command, and needed restarting.
That might be acceptable given our usecase is snapshotting, but still.

I couldn't find any way of testing my RPC call other than live testing. Suggestion from the team was checking in some shell to `scripts/calibnet_health_check.sh`, which seems like more of a smoke test than anything.
If it's true that there's no way to hermetically test our node, then I think we should raise a new issue and add it ASAP - this kind of infrastructure is super important both for feature development, confidence, and learning.

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible)
  - We've added a smoke test in `calibnet_health_check.sh`. It's at the bottom, because this is a dev command that could brick the node.
- [x] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

**Comment by aatifsyed** [comment]
Note, CI has flaked:
https://github.com/ChainSafe/forest/actions/runs/4878069115/jobs/8703365437
```
thread 'tests::bitswap_go_compat_test' panicked at 'called `Result::unwrap()` on an `Err` value: deadline has elapsed', node/forest_libp2p/bitswap/tests/go_compat.rs:27:45
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

   Canceling due to test failure: 0 tests still running
------------
     Summary [  66.565s] 265 tests run: 264 passed, 1 failed, 0 skipped
        FAIL [  65.368s] forest_libp2p_bitswap::go_compat tests::bitswap_go_compat_test
```

I've seen this fail on a local checkout of master too

File: blockchain/chain/src/store/chain_store.rs [hunk 1]
```diff
@@ -306,6 +306,14 @@ where
         })?)
     }
 
+    pub fn unmark_block_as_validated(&self, cid: &Cid) -> Result<(), Error> {
```
Comment (lemmih) [inline blockchain/chain/src/store/chain_store.rs:309]: Conceptually, it's a mistake that the set of validated blocks is saved to a file. But fixing it depends on issue #2771. Once that issue has been closed, this function will be infallible.

File: blockchain/chain/src/store/chain_store.rs [hunk 2]
```diff
@@ -306,6 +306,14 @@ where
...
+        let mut file = self.file_backed_validated_blocks.lock();
+        Ok(file.with_inner(|inner| {
+            // TODO(aatifsyed): should we return an error if block doesn't exist?
```
Comment (lemmih) [inline blockchain/chain/src/store/chain_store.rs:312]: I don't think it's an error if a previously validated block isn't in the set. The set is merely a cache and may be emptied at any point in time.

**Review by lemmih — CHANGES_REQUESTED** [review]
How about testing it against calibnet by setting the head to a few epochs in the past and then verifying that Forest can re-sync to the most recent epoch?

Regarding non-calibnet testing, I am all ears. Let's discuss ways of testing Forest in slack.

Oh, and let's reset the `yarn.lock` file.

File: node/rpc/src/chain_api.rs [hunk 3]
```diff
@@ -323,3 +323,31 @@ where
     let name: String = data.state_manager.chain_config().name.clone();
     Ok(name)
 }
+
+// This is basically a port of the go implementation
```
Comment (elmattic) [inline node/rpc/src/chain_api.rs:327]: Note that there is another go implementation (Venus). Feel free to add in a comment a link to the method/commit that has "inspired" you.

**Review by elmattic — CHANGES_REQUESTED** [review]

File: forest/cli/src/cli/chain_cmd.rs [hunk 4]
```diff
@@ -93,16 +101,46 @@ impl ChainCommands {
...
             Self::ReadObj { cid } => {
-                let cid: Cid = cid.parse()?;
-                print_rpc_res(chain_read_obj((CidJson(cid),), &config.client.rpc_token).await)
+                print_rpc_res(chain_read_obj((CidJson(*cid),), &config.client.rpc_token).await)
             }
+            Self::SetHead {
+                cids,
+                epoch: Some(epoch),
+            } => {
+                assert_eq!(cids.len(), 0, "should be disallowed by clap");
```
Comment (LesnyRumcajs) [inline forest/cli/src/cli/chain_cmd.rs:114]:
```suggestion
                assert!(cids.is_empty(), "should be disallowed by clap");
```

**Review by LesnyRumcajs — COMMENTED** [review]
Looks good, minus test coverage.

**Review by lemmih — APPROVED** [review]

**Review by elmattic — APPROVED** [review]

**Pull request merged**